### DISCUSSION TITLE
Switch to Servomotors for Corners

### DISCUSSION THREAD

**Discussion by ericjunkins** [nasa-jpl/open-source-rover#276] [Ideas]
We've been discussing about potentially overhauling the corner motor system to use Servo motors instead of DC motors, want to create a spot for discussion about this topic. Here are some overall thoughts about this:

## Servo Motor Pros
- Removes Absolute Encoders
- Removes Calibration process associated with Absolute encoders
- Removes 2 Roboclaw motor controllers
- Positioning scheme becomes significantly easier

We are thinking that because of these upsides it is worth doing the effort to design a system to use servo motors instead. This is also going to be in parallel with the PCB Rev F design https://github.com/nasa-jpl/open-source-rover/projects/2 as it requires redoing a lot of electronics.

We're still in early design stages of picking components, and figuring out what will work. Anyone who'd like to participate is more than happy to weigh in thoughts/ideas on this

**Comment by ericjunkins** [comment]
One potential candidate that places nicely with a lot of our system already is the Servocity Servomotor gearbox combo:

https://www.servocity.com/servo-gearbox-642-rotation-700-oz-in-0-32-sec-60/

New Corner system cost:

| Part | qty | cost |
| :----|:---:|:-----:|
| Servo motor | 4 | $120 |
| servo controller | 1 | $15 |

Total: $495

Current Corner system motor cost: 

| Part | qty | cost |
| :----|:---:|:-----:|
| DC motor | 4 | $40 |
| Roboclaw | 2 |  $80 |
| Abs Encoder | 4 | $100 |

Total: $720

Neither of these include mounting hardware, i'm expecting that to be roughly comparable between the two solutions, but can update once a mounting solution is designed

**Reply by alexhx2022** [reply]
I'm a fresh from PCB industry. You can know more about us from [here](https://www.hx-circuit.com/pcb-manufacturer). I want to learn pathon. I don't know why google show me this page. Can I learn pathon here?

**Comment by dcschooley** [comment]
I'm looking at what it would take to drive these. We will need at least one additional voltage regulator. To get 700 oz-in, you need 7.4 volts. Full torque requires 3A, so you would need Pololu's biggest 7.5V regulator ($39.95) to drive all four motors. Another option might be to go up to the 3:1 gear ratio and drive it at 5V to get 720 oz-in @ 2A. A single converter like this: [Pololu 5V converter](https://www.pololu.com/product/4092) could drive all four motors for $19.95 and is somewhat smaller. The available torque goes up to 900 oz-in if you drive the servo at 6V.

**Reply by ericjunkins** [reply]
I think the 700 Oz-in is overkill from what we need, @Achllle I think did a little bit to estimate the torque of the corner system based on current values from the rover.

I gotta think a little bit more on it but in general the torque necessary is also related to the torque of the drive motors. I believe the most strain you'll put on the corners is when you're trying to climb an obstacle, and the back 4 wheels are shoving your front wheel, and the corner has to resist that and hold straight.

**Reply by dcschooley** [reply]
The other place I've seen the corner motors work hard is with soft tires on rough concrete on a hot day or in tall grass. Everything bends enough that you can see the corner wheels get closer together.

**Reply by dcschooley** [reply]
5V on the 700-oz in servos provides a max torque of 480 oz-in. So here's a thought, replace the existing 5V regulator with the big 15A 5V regulator. A single regulator will supply the servos and everything else. If 480 oz-in is enough, that's great. If you want more margin, then the 3:1 gear ratio is an option, and requires less current for normal driving.

**Reply by Achllle** [reply]
Let me get back to you on that, but that sounds good. If I remember correctly, the required torque from experimentation is at most 1.5Nm (~212 oz-in), well under the 480 oz-in. I wrote down the results somewhere but can't remember where. @ericjunkins or @apollokit do any of you happen to remember?

**Reply by dcschooley** [reply]
ServoCity sells replacement gears, so if the initial 2:1 ratio doesn't work, the gears can be replaced for $29 per motor to move to a 3:1 ratio. It sounds like 480 oz-in is more than enough though. When we do our initial testing we can make sure to drive the heck out of the corner motors to see what happens.

**Reply by jeffreyscottgraham** [reply]
Has the accuracy of the new setup been compared?

**Comment by ericjunkins** [comment]
@jeffreyscottgraham going to start a new comment thread to discuss servo accuracy

## Current Setup 

There are also a few factors to consider about our current encoder setup:
- We are using an analog output that ranges from 0-5V, then linearly dividing it to roughly 0-1.6V. This means the _accuracy of our encoder is 0.225 deg_ /mV. 
- Because analog signals have noise (~10-15mV) that corresponds to a 3.375 degree variance based on noise. We use a filtering capacitor to remove high frequency fluctuations in the noise, but all that has done is smooth it, you still obtain that window of inaccuracy 
- Ontop of all that we then have to _calibrate our own PID position controller_ for the motors, which turns out to be not perfect either. This with an underdamped system probably introduces another 1-2 degrees of uncertainty. 

I'm going to be a bad physicist right now and say that these errors are _statistically independent of each other_ and simply add the errors in quadrature. (It won't make that much difference and makes things a lot easier)

![image](https://user-images.githubusercontent.com/25470708/119021420-a61d4d00-b95c-11eb-9717-44e161ebc2c3.png)

for a total uncertainty of `~4 degrees in our current setup`.

Standard servos claim to have `a positional error of around 0.5 deg` 

## Potentiometer vs Encoder

Here's one source ive found talking about a potentiometer vs encoder: [Discussion about potentiometer vs encoder](https://www.midoriamerica.com/news/the-difference-pot-and-encoder/)

To paraphrase: 
 - ``` If you are currently using an encoder with standard 10-bit resolution, its accuracy is roughly the same as the potentiometer with linearity ±0.1%. ``` 
 - Potentiometers are easier to install, as they don't suffer from needing extremely tight alignment as encoders do. This is a win for us, as these are all _hobby level parts_ , you don't have to worry about potential manufacturing issues on signals etc. 



## Conclusion

I `expect with a high level of certainty` that using servos will be more accurate, easier to use, and cheaper. I will of course verify these claims before we finish any of the designs. I currently have the above recommended servo, and am waiting on my servo controller to arrive still. I hope that this alleviates some of your concerns, let me know if you still have anything I didn't address
